optaplanner-core/src/main/java/org/optaplanner/core/impl/util/ConsecutiveSetTree.java

```diff
@@ -0,0 +1,316 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ConsecutiveSetTree<T, I extends Comparable<I>, D extends Comparable<D>> implements Collection<T> {
+    Function<T, I> indexFunction;
+    BiFunction<I, I, D> differenceFunction;
+    D maxDifference;
+    D zeroDifference;
+    TreeMap<T, Sequence<T>> startItemToSequence;
+    Class<? extends T> typeClass;
+    Comparator<T> comparator;
+
+    TreeMapValueList<T, Sequence<T>> sequenceList;
+    ConsecutiveData<T, D> consecutiveData;
+
+    public ConsecutiveSetTree(Class<? extends T> typeClass, Function<T, I> indexFunction,
+            BiFunction<I, I, D> differenceFunction, D maxDifference, D zeroDifference) {
+        this.indexFunction = indexFunction;
+        this.differenceFunction = differenceFunction;
+        this.maxDifference = maxDifference;
+        this.zeroDifference = zeroDifference;
+        this.typeClass = typeClass;
+        // Hashcode for duplicate protection
+        // Ex: two different games on the same time slot
+        comparator = Comparator.comparing(indexFunction).thenComparingInt(System::identityHashCode);
+        startItemToSequence = new TreeMap<>(comparator);
+        consecutiveData = new ConsecutiveData<>(this);
+        sequenceList = new TreeMapValueList<>(startItemToSequence);
+    }
+
+    public List<Sequence<T>> getConsecutiveSequences() {
+        return sequenceList;
+    }
+
+    public List<Break<T, D>> getBreaks() {
+        return startItemToSequence.keySet().stream().flatMap(startItem -> {
+            T nextStartItem = startItemToSequence.higherKey(startItem);
+            if (nextStartItem == null) {
+                return Stream.empty();
+            }
+            T endOfCurrentSequence = startItemToSequence.get(startItem).getItems().last();
+            return Stream.of(new Break<>(nextStartItem, endOfCurrentSequence,
+                    differenceFunction.apply(indexFunction.apply(endOfCurrentSequence),
+                            indexFunction.apply(nextStartItem))));
+        }).collect(Collectors.toList());
+    }
+
+    public List<D> getConsecutiveLengths() {
+        return getConsecutiveSequences().stream()
+                .map(sequence -> differenceFunction.apply(indexFunction.apply(sequence.getItems().first()),
+                        indexFunction.apply(sequence.getItems().last())))
+                .collect(Collectors.toList());
+    }
+
+    public Optional<D> getBreakBefore(Sequence<T> sequence) {
+        T startItem = sequence.getItems().first();
+        T prevStartItem = startItemToSequence.lowerKey(startItem);
+        if (prevStartItem == null) {
+            return Optional.empty();
+        }
+        return Optional
+                .of(differenceFunction.apply(indexFunction.apply(startItemToSequence.get(prevStartItem).getItems().last()),
+                        indexFunction.apply(startItem)));
```

Comment:
This is hard to understand. I would put the individual components into their own variables, making the information flow more obvious.